### PR TITLE
Bug 1269367 - Allow SSL hostname validation in Python 2.7.9+

### PR DESCRIPTION
**Pull request by gregarndt**
Socket connections can be made to an IP address instead of a hostname,
but httplib connections should maintain the hostname in self.host for
SSL certificate validation.

### CODE REVIEW

File: storage.py [hunk 1]
```diff
@@ -271,33 +272,39 @@ def ConnectionWrapperFactory(parent_class, dns_query):
...
+        def create_connection(self, *args, **kwargs):
+            '''
+            httplib.HTTP(S)Connection creates a socket connection to
+            self.host using _create_connection, which by default is a
+            redirect to socket.create_connection(). Connections are made to
+            self.host by default.  When attempting to perform DNS resolution
+            and connecting to an IP address instead of hostname, SSL certificate
+            validation is broken in Python 2.7.9+.  This wrapper allows for
+            a dns query to be made, and socket connection made to an IP address
+            whild still maintaining self.host as the unresolved hostname for
```
Comment (glandium) [inline storage.py]: typo here

File: storage.py [hunk 2]
```diff
@@ -271,33 +272,39 @@ def ConnectionWrapperFactory(parent_class, dns_query):
     method uses the dns_query function to resolve the connection host name.
     '''
     class ConnectionWrapper(parent_class):
-        def connect(self):
+        def create_connection(self, *args, **kwargs):
+            '''
+            httplib.HTTP(S)Connection creates a socket connection to
+            self.host using _create_connection, which by default is a
+            redirect to socket.create_connection(). Connections are made to
```
Comment (glandium) [inline storage.py]: It might be worth mentioning that self._create_connection exists for python unit tests.

File: storage.py [hunk 3]
```diff
@@ -271,33 +272,39 @@ def ConnectionWrapperFactory(parent_class, dns_query):
...
-            # we can change self.host for use when opening the socket. However,
-            # boto reuses HTTP(S)Connection instances, and on the next request,
-            # httplib resets the Host header with self.host, so we need to
-            # restore it.
-            host = self.host
-            self.host = dns_query(self.host)
+            resolvedHost = dns_query(self.host)
             t1 = time.time()
             _last_stats['dns'] = (t1 - t0) * 1000
+            sock = socket.create_connection(*args, **kwargs)
```
Comment (glandium) [inline storage.py]: without actually modifying the arguments, socket.create_connection is going to do its own dns resolution. If the goal was to measure the time it takes to do a dns resolution and have the socket creation do its own, we wouldn't have this wrapping in the first place ;). So you need to actually modify the arguments to replace the host with the resolved host.

File: storage.py [hunk 4]
```diff
@@ -271,33 +272,39 @@ def ConnectionWrapperFactory(parent_class, dns_query):
...
+            _last_stats['socket_connect'] = time.time()
```
Comment (glandium) [inline storage.py]: _last_stats is supposed to only contain durations, not timestamps. Plus, the starting point for the "connect" duration should still be t1, before create_connection is called.

File: storage.py [hunk 5]
```diff
@@ -271,33 +272,39 @@ def ConnectionWrapperFactory(parent_class, dns_query):
...
 
         def getresponse(self, buffering=False):
             res = parent_class.getresponse(self, buffering)
             _last_stats['response'] = (time.time() - self._connect_time) * 1000
             return res
 
-
     return ConnectionWrapper
 
-
```
Comment (glandium) [inline storage.py]: This line should stay (PEP8)

File: storage.py (hunk 3 above)
Comment (gregarndt) [inline storage.py]: Ah very good point! I did that at first, and then for some slip of the mind I changed it to this mistakenly.  Good catch!

**Comment by gregarndt** [comment]
Thanks for the comments! I addressed them (I hope).

**Comment by glandium** [comment]
Please test the following patch on top of your PR. If it works, I'll fold it all and push to master.
```diff
diff --git a/storage.py b/storage.py
index 6ef7c9a..d952b88 100644
--- a/storage.py
+++ b/storage.py
@@ -272,7 +272,7 @@ def ConnectionWrapperFactory(parent_class, dns_query):
     method uses the dns_query function to resolve the connection host name.
     '''
     class ConnectionWrapper(parent_class):
-        def create_connection(self, *args, **kwargs):
+        def create_connection(self, address, timeout, source_address):
             '''
             httplib.HTTP(S)Connection creates a socket connection to
             self.host using _create_connection, which by default is a
@@ -286,18 +286,20 @@ def ConnectionWrapperFactory(parent_class, dns_query):
             as the unresolved hostname for validation.
             '''
             t0 = time.time()
-            resolvedHost = dns_query(self.host)
-            self._socket_connect_timestamp = time.time()
-            _last_stats['dns'] = (self._socket_connect_timestamp - t0) * 1000
-            sock = socket.create_connection((resolvedHost,self.port),
-                                           self.timeout, self.source_address)
+            host, port = address
+            resolvedHost = dns_query(host)
+            self._socket_connect_timestamp = t1 = time.time()
+            _last_stats['dns'] = (t1 - t0) * 1000
+            sock = socket.create_connection((resolvedHost, port),
+                                            timeout, source_address)
             return sock
 
         def connect(self):
             self._create_connection = self.create_connection
             parent_class.connect(self)
             self._connect_time = time.time()
-            _last_stats['connect'] = (self._connect_time - self._socket_connect_timestamp) * 1000
+            _last_stats['connect'] = (
+                self._connect_time - self._socket_connect_timestamp) * 1000
 
         def getresponse(self, buffering=False):
             res = parent_class.getresponse(self, buffering)
```

**Comment by gregarndt** [comment]
Thanks for the suggested changes, sorry for the roundabout on this.  I have applied the changes in  	 5908a92, and was still able to put/get a file into s3 by running:

```python
from storage import Storage
from time import time

fn = "test{}".format(time())

storage = Storage.from_environment()
storage.put(fn, 'test')
assert storage.get(fn) == 'test'
```

**Comment by glandium** [comment]
Merged in e3f5fb2

**Pull request closed without merging**